core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+        fragments.forEach(fragment -> visitor.processFragment(fragment.getId()));
+
+        // Make sure there are no strongly connected components as it would mean circular dependency between stages
+        List<Set<PlanFragmentId>> components = new StrongConnectivityInspector<>(fragmentDependency).stronglyConnectedSets();
+        verify(components.stream().allMatch(set -> set.size() == 1), "circular dependency between stages");
+
+        return nonLazyFragments;
+    }
+
+    private static class Visitor
+            extends PlanVisitor<Fragments, PlanFragmentId>
+    {
+        private final Map<PlanFragmentId, PlanFragment> fragments;
+        private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentDependency;
+        private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentTopology;
+        private final Set<PlanFragmentId> nonLazyFragments;
+        private final Map<PlanFragmentId, Fragments> fragmentSources = new HashMap<>();
+
+        public Visitor(
+                Collection<PlanFragment> fragments,
+                DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentDependency,
+                DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentTopology,
+                Set<PlanFragmentId> nonLazyFragments)
+        {
+            this.fragments = requireNonNull(fragments, "fragments is null").stream()
+                    .collect(toImmutableMap(PlanFragment::getId, identity()));
+            this.fragmentDependency = requireNonNull(fragmentDependency, "fragmentDependency is null");
+            this.fragmentTopology = requireNonNull(fragmentTopology, "fragmentTopology is null");
+            this.nonLazyFragments = requireNonNull(nonLazyFragments, "nonLazyFragments is null");
+        }
+
+        public Fragments processFragment(PlanFragmentId planFragmentId)
+        {
+            if (fragmentSources.containsKey(planFragmentId)) {
+                return fragmentSources.get(planFragmentId);
+            }
+
+            Fragments fragment = processFragment(fragments.get(planFragmentId));
+            fragmentSources.put(planFragmentId, fragment);
+            return fragment;
+        }
+
+        private Fragments processFragment(PlanFragment fragment)
+        {
+            Fragments source = fragment.getRoot().accept(this, fragment.getId());
+            // append current fragment to set of all fragments
+            Set<PlanFragmentId> allFragments = ImmutableSet.<PlanFragmentId>builder()
+                    .addAll(source.getAllFragments())
+                    .add(fragment.getId())
+                    .build();
+            Set<PlanFragmentId> lazyFragments;
+            if (source.isCurrentFragmentLazy()) {
+                // append current fragment as a lazy fragment
+                lazyFragments = ImmutableSet.<PlanFragmentId>builder()
+                        .addAll(source.getLazyFragments())
```

Comment:
Declare `lazyFragments = ImmutableSet.<PlanFragmentId>builder()` before `if` and also move adding `source.getLazyFragments();` to it before `if`